Xml2star conversion particle path bug

McHaillet opened the pull request:
Pytom writes particle annotions in .xml format, but for compatibility with Relion and Warp it needs to be converted to a .star file format. Pytom already supported conversion to a simplistic pickpostion format that could be read by warp for subtomogram reconstruction. I now added support to allow back and forth conversion with star files that refer to subtomograms and ctf volumes that Relion needs for subtomogram averaging and classification.

- this prepares pytom for some subtomogram averaging update I wanted to add
- and closes #126

I updated unittests to make sure the old behaviour is consistent, and the back and forth conversion with the updated format is consistent (no information loss).

sroet commented on pytom/basic/datatypes.py:
```diff
@@ -236,38 +236,34 @@ def generate_default_alignmentresults(folder, metafilename='', prefix='sorted_',
-                     ('MaxValueProbDistribution', 'f4'),
-                     ('NrOfSignificantSamples', 'i4'),
-                     ('RandomSubset', 'i4')
-                     ]
+                              ('CoordinateY', 'f4'),
+                              ('CoordinateZ', 'f4'),
+                              ('MicrographName', 'U1000'),
+                              ('Magnification', 'f4'),
+                              ('DetectorPixelSize', 'f4'),
+                              ('GroupNumber', 'i4'),
+                              ('AngleRot', 'f4'),
+                              ('AngleTilt', 'f4'),
+                              ('AnglePsi', 'f4'),
+                              ('CtfMaxResolution', 'f4'),
+                              ('ImageName', 'U1000'),
+                              ('CtfImage', 'U1000'),
+                              ('PixelSize', 'f4'),
+                              ('OriginXAngst', 'f4'),
+                              ('OriginYAngst', 'f4'),
+                              ('OriginZAngst', 'f4'),
+                              ('ClassNumber', 'i4'),
+                              ('Voltage', 'f4'),
+                              ('SphericalAberration', 'f4')]
 
```
Any reason why one of these is not a subset of the other? You both add and remove data-types, why?

McHaillet replied:
Good that you ask

Well, this format (DATATYPE_RELION31_EXTENDED) is only used when converting a pytom particle list (xml) to a Relion/Warp particle list (star). Its not used when reading star files! I.e. pytom is able to read a lot more extensive star files. Also the format was not used before this commit to write star files (it was only used in a unittest). But actually will double check this test for the behaviour

The columns I changed for the output are to ensure that Relion3 can read it without issue. I used the Warp star file header as an example, because I know Relion can read those.

Some details on the changes:

removed
('NormCorrection', 'f4'),
('LogLikeliContribution', 'f4'),
('MaxValueProbDistribution', 'f4'),
('NrOfSignificantSamples', 'i4'),
('RandomSubset', 'i4')

These values are from the Relion optimisation process. They cannot be stored in pytom's particle list format and are particle specific. So it is not possible to supply them by the user when converting xml to stars. Relion also does not need them to read the particle list, so the data can be transferred to Relion without issue. For new optimisation processes in Relion they anyway need to be estimated anew.

removed
('OpticsGroup', 'i4')

Opticsgroups in Relion were introduced for high-order aberration corrections. For subtomogram averaging from Warp this info is useless because Relion does not have the original micrographs (and a fixed ctf model). So there is nothing to optimise there. Relion can read it without issue if this is not present.

added
('DetectorPixelSize', 'f4')

Added because Warp also has it.

added
('Magnification', 'f4')
('Voltage', 'f4'),
('SphericalAberration', 'f4')

Added because Relion gives an error message if they are not present.

sroet reviewed (changes requested):
One comment about a strange format change, and I would very much also like unittest for the newly added fail paths (so the code should fail with the expected error if it is asking with tomo data without the new keywords)

McHaillet commented:
> and I would very much also like unittest for the newly added fail paths (so the code should fail with the expected error if it is asking with tomo data without the new keywords)

Good point, will do!

McHaillet commented:
I updated as requested:

 - added tests that check for error raising by pl2star for missing spherical_aberration or voltage

And also changed the following:

 - removed ERROR text from pl2star because exception should already be informative enough. It also made the unittest confusing.
 - updated the IOTest to not read an explicit relion star header, star file reading/writing should be able to read any starfile and write any numpy structured array to a star

sroet reviewed (approved):
LGTM, feel free to merge
